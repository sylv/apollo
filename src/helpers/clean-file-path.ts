import path from "path";
import { ALL_EXTENSIONS, SUBTITLE_EXTENSIONS } from "../constants";
import { stripSpacePlaceholders } from "./strip-space-placeholders";

const EXPLICITLY_STRIP_PARTS = new Set([
  "tv shows",
  "videos",
  "movies",
  "shows",
  "series",
  "seasons",
  "import",
  "clean",
  "data",
  "library",
  "media",
  "mnt",
  "downloads",
  "active",
  "complete",
  "incomplete",
  "downloading",
  "torrents",
  "torrent",
]);

const IGNORE_PATH_PART_REGEX = /^(?:[A-Z]:|[a-z]+)$/i;
const EXCLUDE_BLACKLIST_REGEX = [
  /^lore$/i,
  /^histories(( and| &) lore)?$/i,
  /sample|^trailer(?! park)/i, // "excludes "trailer park boys"
  /^behind.the.scenes$/i,
  /^deleted.and.extended.scenes$/i,
  /^deleted.scenes$/i,
  /^extras?$/i,
  /^featurettes$/i,
  /^interviews$/i,
  /^scenes$/i,
  /^shorts$/i,
];

/**
 * Clean the file path and replace dots or underscores with spaces when necessary.
 * @returns undefined if the path is blacklisted.
 */
export function cleanFilePath(filePath: string): string | undefined {
  const filePathParts = filePath.split(/\/|\\/g);
  const cleanedPathParts: string[] = [];
  let hitUnignoredPart = false;

  for (let i = 0; i < filePathParts.length; i++) {
    const filePathPart = filePathParts[i];
    const lower = filePathPart.toLowerCase();
    if (EXPLICITLY_STRIP_PARTS.has(lower)) {
      continue;
    }

    const isLastPart = i === filePathParts.length - 1;
    const ext = isLastPart ? ALL_EXTENSIONS.find((ext) => lower.endsWith(ext)) : null;
    const withoutExtension = ext ? filePathPart.slice(0, -ext.length) : filePathPart;
    const cleanPathPartWithoutExt = stripSpacePlaceholders(withoutExtension);
    const cleanPathPart = ext ? `${cleanPathPartWithoutExt}${ext}` : cleanPathPartWithoutExt;
    // we have to check the blacklist first or else something like "Trailers" would be discarded
    // before we check the blacklist
    if (EXCLUDE_BLACKLIST_REGEX.some((p) => p.test(cleanPathPart))) return;
    // only once it passes the blacklist can we do this check
    if (filePathPart.length <= 2 || (IGNORE_PATH_PART_REGEX.test(filePathPart) && !hitUnignoredPart)) continue;
    hitUnignoredPart = true;
    cleanedPathParts.push(cleanPathPart);
  }

  return cleanedPathParts.join("/");
}
