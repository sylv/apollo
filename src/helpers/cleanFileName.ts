import { fileExtensionRegex } from '../constants';
import { stripNameTags } from './stripNameTags';

const groupExtra = /[\-]{1,}[A-z0-9]{2,}$/;
const randomShit = /dts|(dd)?((5|7)\.1(?:ch)?|ch|Atmos|nf|YIFY|Tigole|multi|hc|Dolby|TrueHD|4k|HDR|Blu-?Ray|final cut|web-?[A-z]+|bd-?rip|dvd-?rip|UHD|ReEnc|(?:kor)?sub|[0-9]{1,2}bit)/;
const excessRegex = new RegExp(`(${groupExtra.source}|(\\[|\\(|\\.|-| )?${randomShit.source}(?:\\)|\\]|\\.|-| |$))`, 'ig');
const specialDoubleSpaceRegex = /[ .]{2,}/g;
const cleanTagEndRegex = / ?-? (\]|\))/g;
const cleanTagStartRegex = /(\[|\() ?-? /g;

/**
 * Convert a filthy torrent name to sommething usable.
 * - Removes junk we don't need that could get in the way, like "TrueHD" and shit like that.
 * - Replaces full stops with spaces when the torrent is likely using them in place of spaces
 * - Replaces underscores with spaces when the torrent is likely using themm in place of spaces
 * - Removes double-spaces, "  " => " "
 */
export function cleanFileName(fileName: string) {
  const extensionMatch = fileName.match(fileExtensionRegex);
  const extension = (extensionMatch && extensionMatch[0]) || undefined;
  if (extension) {
    fileName = fileName.slice(0, -extension.length);
  }

  // strip some useless things without damaging other tags
  let cleanedFileName = stripNameTags(fileName, excessRegex);

  // some files use full stops instead of spaces
  // some files use underscores instead of spaces
  // some files use spaces like they should be used
  // and some torrents use primarily full stops with some spaces on the end...
  // and some torrents have full stops in the title but use spaces.
  // and some torrents have underscores in some places but use primarily spaces.
  // we'll convert to spaces only if it looks like the title is using mostly full stops or underscores
  // we intentionally use fileName in place of cleanedFileName as preprocessing might have introduced more spaces then full stops
  // we also give spaces a bit of a handicap because when used properly, full stops should be used minimally
  const spaceCount = fileName.split(' ').length / 1.4;
  const underscoreCount = fileName.split('_').length;
  const fullStopCount = fileName.split('.').length;
  const hasMoreUnderscores = underscoreCount > spaceCount && underscoreCount > fullStopCount;
  const hasMoreFullStops = fullStopCount > spaceCount && fullStopCount > underscoreCount;
  if (hasMoreFullStops) {
    cleanedFileName = cleanedFileName.split('.').join(' ');
  } else if (hasMoreUnderscores) {
    cleanedFileName = cleanedFileName.split('_').join(' ');
  }

  cleanedFileName = cleanedFileName
    // remove double spaces
    // also threw in the . to fix "Avengers Endgame 2019. 2160p. Ita Eng x265" handling
    // not sure about ^ because it kinda undoes a lot of the work we did above.
    .replace(specialDoubleSpaceRegex, ' ')
    // if we strip part of a tag it could leave remenants, e.g `(1080p - )`
    // these two replaces will turn that into `(1080p)
    .replace(cleanTagEndRegex, '$1')
    .replace(cleanTagStartRegex, '$1')
    .trim();

  return {
    extension,
    cleanedFileName
  };
}
