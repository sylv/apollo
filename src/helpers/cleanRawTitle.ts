import { titleCase } from "title-case";
import { IGNORE_TITLES } from "../data";

const TITLE_TAG_REGEX = /\[.*?\]/g;
const TITLE_RELEASE_GROUP_REGEX = /-[a-z]{2,}(?=$|\/)/gi;
const TITLE_TRAILING_TAG_REGEX = / ?(?:\(|\[|\-) ?$/g;
const TITLE_URL_REGEX = /(www(?:\.| ))?[a-z0-9]+(?:\.| )(?:com|org|me|se|info)/i;
const TITLE_BRACKETS_PREFIX_REGEX = /^\([A-z0-9]+\) ?/;
const PREFIX_REGEX = /^ ?(\[|\]|\(|\)|[0-9]-|-|\/)/g;
const SUFFIX_REGEX = /(\[|\]|\(|\)|-|\.|\/) ?$/g;
const SPECIAL_CHARACTER_REGEX = /^[\W ]+$/;

/**
 * strip information that was incorrectly included in a title.
 * @example "(auto) Infinity War Movie 1 (" -> "Infinity War"
 */
export function cleanRawTitle(title: string) {
  const clean = title
    // remove tags like [1080p]
    .replace(TITLE_TAG_REGEX, " ")
    // remove release group suffixes like -QxR
    .replace(TITLE_RELEASE_GROUP_REGEX, "")
    // trim trailing ('s and -'s from e.g "Avatar - The Last Airbender Movie ("
    .replace(TITLE_TRAILING_TAG_REGEX, "")
    // remove urls that weren't in brackets
    .replace(TITLE_URL_REGEX, "")
    // remove things like "(auto)" at the start of strings, e.g "(auto) Top Gear"
    .replace(TITLE_BRACKETS_PREFIX_REGEX, "")
    // strip "1-" in "1-Fellowship of the Ring"
    // strip the "[" in "[ Title" etc
    .replace(PREFIX_REGEX, "")
    .replace(SUFFIX_REGEX, "")
    .trim();

  // if after cleaning the title is empty or is in IGNORE, it's useless
  if (!clean || IGNORE_TITLES.has(clean.toLowerCase())) return;

  // the commented code below breaks on something like "Rick and Morty/Season 5/Episode 3"
  // and I can't think of anything to prevent that so fuck it let's just not.
  // // if the title includes path separators it's likely an invalid match.
  // if (clean.includes("/")) return;

  // if the entire string is special characters, we probably failed
  // to extract a valid title.
  if (SPECIAL_CHARACTER_REGEX.test(clean)) return;
  return titleCase(clean);
}
