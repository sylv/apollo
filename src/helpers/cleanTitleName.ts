import { tagStart, tagEnd } from './stripNameTags';
import { titleCase } from './titleCase';

const cleanTitleNameRegex = /^( |-|,|\]|\[|\)|-)+|( |-|,|\]|\[|\(|-)+$/g;

/**
 * Remove tags at the start of a title. Some titles end with useful tags at the end, so we preserve those - ex "Avatar (tlok)"
 */
function removeStartTags(input: string) {
  let sliceTo = 0;
  let inTag = false;
  for (let i = 0; i < input.length; i++) {
    const char = input[i];
    if (char.trim() === '') continue;
    if (tagStart.includes(char)) {
      inTag = true;
      continue;
    }

    if (tagEnd.includes(char)) {
      inTag = false;
      sliceTo = i;
      continue;
    }

    if (inTag) {
      continue;
    }

    break;
  }

  if (sliceTo !== 0) {
    return input.substring(sliceTo + 1).trim();
  }

  return input.trim();
}

/**
 * Clean a title name. "avengers infinity war-" => "Avengers Infinity War"
 * @param input The dirty title. If the title already contains capitalisation we'll leave it alone.
 */
export function cleanTitleName(rawInput: string) {
  // if the title starts with a tag, chances are it's worthless garbage. But if it contains a random tag, it could be an abbreviation.
  // specifically this is the handle "Avatar (TLA)" while also stripping the useless tag from "[pseudo] My Show S01E01"
  const withoutStartTags = removeStartTags(rawInput);
  const withoutGarbage = withoutStartTags.replace(cleanTitleNameRegex, ' ');
  return titleCase(withoutGarbage);
}
