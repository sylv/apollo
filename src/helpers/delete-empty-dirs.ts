import { readdir, rmdir } from "fs/promises";
import { join } from "path";

const isDeletable = async (directory: string) => {
  const entries = await readdir(directory, { withFileTypes: true });
  for (const entry of entries) {
    const entryPath = join(directory, entry.name);
    if (entry.isDirectory()) {
      const shouldDelete = await isDeletable(entryPath);
      if (!shouldDelete) {
        // child directory has files so the child *and* parent cannot be deleted
        return false;
      }

      await rmdir(entryPath);
    } else {
      return false;
    }
  }

  return true;
};

/**
 * Delete empty directories in the given directory.
 * @param directory The directory to scan, this directory will not be deleted if empty after this function is called.
 */
export const deleteEmptyDirs = async (directory: string) => {
  // todo: should use opendir() but its not available in tests
  const entries = await readdir(directory, { withFileTypes: true });
  for (const entry of entries) {
    if (entry.isDirectory()) {
      const entryPath = join(directory, entry.name);
      const shouldDelete = await isDeletable(entryPath);
      if (shouldDelete) {
        await rmdir(entryPath);
      }
    }
  }
};
