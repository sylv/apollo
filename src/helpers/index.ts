export * from "./clean-file-path";
export * from "./clean-raw-title";
export * from "./get-all-matches";
export * from "./strip-space-placeholders";
