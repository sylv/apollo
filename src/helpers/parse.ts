import { ApolloParser } from "../classes/apollo-parser";

export const parse = (input: string, providers = ["local", "imdb"]) => {
  const parser = new ApolloParser({ providers });
  return parser.parse(input);
};
