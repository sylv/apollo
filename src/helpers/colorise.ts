import chalk from "chalk";
import dedent from "dedent";
import { inspect } from "util";
import { ApolloParser } from "../classes/apollo-parser";
import { ApolloOutput } from "../types";
import { cleanFilePath } from "./clean-file-path";

const useColours = ["blue", "magenta", "cyan", "gray", "green", "yellow", "red"] as const;
const resolved = new Map<string, string>();
let offset = 0;

const getChalkPropertyColour = (input: string) => {
  const existing = resolved.get(input);
  if (existing) {
    return {
      wrap: (chalk as any)[existing],
      seen: true,
    };
  }

  const nextColour = useColours[offset++ % useColours.length];
  resolved.set(input, nextColour);
  return {
    wrap: (chalk as any)[nextColour],
    seen: false,
  };
};

const filterKeys = (input: Record<string, any>) => {
  const filtered: Record<string, any> = {};
  for (const [key, value] of Object.entries(input)) {
    if (value === undefined || value === null) continue;
    if (Array.isArray(value) && value.length === 0) continue;
    if (typeof value === "object" && !Array.isArray(value)) {
      const result = filterKeys(value);
      if (Object.keys(result).length !== 0) {
        filtered[key] = result;
      }

      continue;
    }

    filtered[key] = value;
  }

  return filtered;
};

export const coloriseResult = (input: string, result: ApolloOutput, parser: ApolloParser) => {
  const cleanInput = cleanFilePath(input);
  const matches = parser.matchIndexes;
  if (!cleanInput) {
    throw new Error(`Invalid input "${input}"`);
  }

  let colorised = "";
  const colourKeys: string[] = [];
  for (let charIndex = 0; charIndex < cleanInput.length; charIndex++) {
    const char = cleanInput[charIndex];
    const matchContainingChar = matches.find((match) => match.start <= charIndex && charIndex < match.end);
    if (!matchContainingChar) {
      colorised += char;
      continue;
    }

    const propertyName = matchContainingChar.propertyName ?? "unknown extractor";
    const propertyColour = getChalkPropertyColour(propertyName);
    colorised += propertyColour.wrap(char);
    if (!propertyColour.seen) {
      colourKeys.push(propertyColour.wrap(propertyName));
    }
  }

  const inspected = inspect(filterKeys(result), { colors: true, depth: null });

  // avert your eyes.
  // dedent doesnt like inspected results and breaks, so it has to be added separately.
  return (
    "\n\n" +
    dedent`
    ${chalk.bold("Input")}
    ${colorised}

    ${chalk.bold("Extracted properties")}
    ${colourKeys.join(", ")}
  ` +
    `\n
${chalk.bold("Result")}
${inspected}` +
    "\n\n"
  );
};
