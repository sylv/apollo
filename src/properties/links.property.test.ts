import { parse } from "../helpers/parse";

it("should extract ids", () => {
  expect(parse("[id=t3_vavr7h]")).toMatchSnapshot();
  expect(parse("1525508539468419074_1")).toMatchSnapshot();
  expect(parse("1525508539468419074_1")).toMatchSnapshot();
  expect(parse("test-ph5d48409d154c4")).toMatchSnapshot();
  expect(parse("https://www.pornhub.com/view_video.php?viewkey=ph5d48409d154c4")).toMatchSnapshot();
});
