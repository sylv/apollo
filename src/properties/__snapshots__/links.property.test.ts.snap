// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should extract ids 1`] = `Promise {}`;

exports[`should extract ids 2`] = `Promise {}`;

exports[`should extract ids 3`] = `Promise {}`;

exports[`should extract ids 4`] = `Promise {}`;

exports[`should extract ids 5`] = `Promise {}`;
